daemon/oci_linux.go

```diff
@@ -732,14 +732,27 @@ func (daemon *Daemon) createSpec(c *container.Container) (*specs.Spec, error) {
 		return nil, err
 	}
 
-	if err := daemon.setupSecretDir(c); err != nil {
+	var hasSecretDir bool
+	defer func() {
+		if hasSecretDir && err != nil {
+			daemon.cleanupSecretDir(c)
+		}
+	}()
+
+	if err := daemon.setupSecretDir(c, &hasSecretDir); err != nil {
```

Comment:
I think it would make the code a bit messier, since the function would have to take a boolean argument and return a new value. The calling code would need logic to act on the returned value, and forgetting to do this would lead to errors. But I'm fine with making the change if you prefer it.

The other thing we could consider is moving `hasSecretDir` to `Daemon`, which might be cleaner overall.